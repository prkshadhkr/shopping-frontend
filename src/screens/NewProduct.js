import React from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { addProductAPI } from '../action/products';
import ProductForm from './ProductForm';

const NewProduct = () => {
  const history = useHistory();
  const dispatch = useDispatch();

  /** add product and save to backend */
  const save = ({
    name,
    image, 
    brand,
    price,
    category,
    count_in_stock,
    description
  }) => {
    dispatch( addProductAPI(
      name,
      image,
      brand,
      +price,
      category,
      +count_in_stock,
      description
    ));
    
    history.push("/products")
  }

  /** cancel and redirect */
  const cancel = () => {
    history.push("/");
  }
  
  return (
    <main>
      <h5 style={{textAlign: "center"}}>New Product</h5>
      <ProductForm save={save} cancel={cancel} />
    </main>
  );
}

export default NewProduct;
