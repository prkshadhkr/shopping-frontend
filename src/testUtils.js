import React from  'react';
import UserContext from './user/UserContext';

/** following is for mock store */
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter';

const mock = new MockAdapter(axios);
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {};
const makeMockStore = (state = {}) => {
  return mockStore ({
    ...initialState,
    ...state
  })
}

/** create a demouser */
const demoUser = {
  username: "testing",
  first_name: "testing",
  last_name: "testing",
  email: "testing@gmail.com",
  password: "testing",
  is_admin: false
};

/**
 * user is destructured from userContext in Profilejs as
 * const { user, setUser } = useContext(UserContext);
 */
const UserProvider = ({ children, user = demoUser }) => (
  <UserContext.Provider value={{user}}>
    {children}
  </UserContext.Provider>
)

export { mock, makeMockStore, UserProvider };