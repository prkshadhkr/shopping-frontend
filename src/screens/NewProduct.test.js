import React from 'react';
import { render } from '@testing-library/react';
import NewProduct from './NewProduct';
import { MemoryRouter } from 'react-router';
import { Provider } from "react-redux";
import { makeMockStore } from '../testUtils';

const store = makeMockStore();

it ('renders without crashing', function() {
  render(
    <MemoryRouter>
      <Provider store={store}>
        <NewProduct />
      </Provider>
    </MemoryRouter>
  )
});

it ('matches snapshot', function() {
  const { asFragment } = render (
    <MemoryRouter>
      <Provider store={store}>
        <NewProduct />
      </Provider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
})