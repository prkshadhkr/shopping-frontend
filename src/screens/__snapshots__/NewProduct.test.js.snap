// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <main>
    <h5
      style="text-align: center;"
    >
      New Product
    </h5>
    <form
      class="mb-4"
    >
      <div
        class="form-group row"
      >
        <label
          class="col-sm-2 col-form-label"
          for="for-name"
        >
          Name
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="for-name"
            name="name"
            value=""
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-2 col-form-label"
          for="for-image"
        >
          Image
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="for-image"
            name="image"
            value=""
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-2 col-form-label"
          for="for-brand"
        >
          Brand
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="for-brand"
            name="brand"
            value=""
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-2 col-form-label"
          for="for-price"
        >
          Price
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="for-price"
            name="price"
            type="number"
            value=""
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-2 col-form-label"
          for="for-category"
        >
          Category
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="for-category"
            name="category"
            value=""
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-2 col-form-label"
          for="for-count_in_stock"
        >
          In stock
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="for-count_in_stock"
            name="count_in_stock"
            type="number"
            value=""
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-sm-2 col-form-label"
          for="for-description"
        >
          Description
        </label>
        <div
          class="col-sm-10"
        >
          <textarea
            class="form-control"
            id="for-description"
            name="description"
          />
        </div>
      </div>
      <button
        class="btn btn-outline-primary float-right"
      >
         Save 
      </button>
      <button
        class="btn btn-outline-danger mr-2 float-right "
      >
         Cancel 
      </button>
    </form>
  </main>
</DocumentFragment>
`;
